vsphere/internal/virtualdevice/virtual_machine_cdrom_subresource.go

```diff
@@ -389,30 +414,40 @@ func (r *CdromSubresource) Create(l object.VirtualDeviceList) ([]types.BaseVirtu
 	}
 
 	// We now have the controller on which we can create our device on.
-	dsID := r.Get("datastore_id").(string)
-	path := r.Get("path").(string)
-	ds, err := datastore.FromID(r.client, dsID)
-	if err != nil {
-		return nil, fmt.Errorf("cannot find datastore: %s", err)
-	}
-	dsProps, err := datastore.Properties(ds)
+	err = r.ValidateCdromConfig()
 	if err != nil {
-		return nil, fmt.Errorf("could not get properties for datastore: %s", err)
-	}
-	dsName := dsProps.Name
-
-	dsPath := &object.DatastorePath{
-		Datastore: dsName,
-		Path:      path,
+		return nil, err
 	}
-
+	dsID := r.Get("datastore_id").(string)
+	path := r.Get("path").(string)
+	clientDevice := r.Get("client_device").(bool)
 	device, err := l.CreateCdrom(ctlr.(*types.VirtualIDEController))
 	if err != nil {
 		return nil, err
 	}
-	device = l.InsertIso(device, dsPath.String())
-	l.Connect(device)
-
+	if dsID != "" && path != "" {
+		// If the datastore ID and path are both set, the CDROM will be backed by a file on a datastore
+		ds, err := datastore.FromID(r.client, dsID)
+		if err != nil {
+			return nil, fmt.Errorf("cannot find datastore: %s", err)
+		}
+		dsProps, err := datastore.Properties(ds)
+		if err != nil {
+			return nil, fmt.Errorf("could not get properties for datastore: %s", err)
+		}
+		dsName := dsProps.Name
+		dsPath := &object.DatastorePath{
+			Datastore: dsName,
+			Path:      path,
+		}
+		device = l.InsertIso(device, dsPath.String())
+		l.Connect(device)
+	} else if clientDevice {
```

Comment:
Let's try to avoid `else if` as it's not necessarily good Go.

This actually can be written as a switch:

```go
switch {
case dsID != "" && path != "":
  ...
case clientDevice == true:
  // You might want to use the full-form boolean expression above to be specific about intent
  ...
default:
  // This would be where your "else" goes, if you had one
}
```